library/common/include/c_types.h

```diff
@@ -0,0 +1,147 @@
+#pragma once
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+// NOLINT(namespace-envoy)
+
+/**
+ * Handle to an Envoy engine instance. Valid only for the lifetime of the engine and not intended
+ * for any external interpretation or use.
+ */
+typedef uint64_t envoy_engine_t;
+
+/**
+ * Handle to an outstanding Envoy HTTP stream. Valid only for the duration of the stream and not
+ * intended for any external interpretation or use.
+ */
+typedef uint64_t envoy_stream_t;
+
+/**
+ * Result codes returned by all calls made to this interface.
+ */
+typedef enum { ENVOY_SUCCESS, ENVOY_FAILURE } envoy_status_t;
+
+/**
+ * Error code associated with terminal status of a HTTP stream.
+ */
+typedef enum { ENVOY_STREAM_RESET } envoy_error_code_t;
+
+/**
+ * Common abstraction for strings.
+ */
+typedef struct {
+  uint64_t length;
+  const char* data;
+} envoy_string;
+
+/**
+ * Holds raw binary data as an array of bytes.
+ */
+typedef struct {
+  uint64_t length;
+  uint8_t* bytes;
+} envoy_data;
+
+/**
+ * Holds data about an HTTP stream.
+ */
+typedef struct {
+  // Status of the Envoy HTTP stream. Note that the stream might have failed inline.
+  // Thus the status should be checked before pursuing other operations on the stream.
+  envoy_status_t status;
+  // Handle to the Envoy HTTP Stream.
+  envoy_stream_t stream;
+} envoy_stream;
+
+/**
+ * Holds a single key/value header.
+ */
+typedef struct {
+  envoy_string key;
+  // Multiple header values for the same header key are supported via a comma-delimited string.
+  envoy_string value;
```

Comment:
Personally I would swap with envoy_data here as a general indication that these things are not necessarily null terminated.